### PR TITLE
# [HINTS]Add implementation for hints on functions fast_ec_add and ec_mul_inner (cairo_secp/ec.cairo)

### PR DESCRIPTION
**Pull request by pefontana**
# [HINTS]Add implementation for hints on functions fast_ec_add and ec_mul_inner (cairo_secp/ec.cairo)

## Description

[cairo_secp/ec.cairo](https://github.com/starkware-libs/cairo-lang/blob/167b28bcd940fd25ea3816204fa882a0b0a49603/src/starkware/cairo/common/cairo_secp/ec.cairo#L188)

* Hints implementation:

```
fast_ec_add_assign_new_x:
%{
    from starkware.cairo.common.cairo_secp.secp_utils import SECP_P, pack

    slope = pack(ids.slope, PRIME)
    x0 = pack(ids.point0.x, PRIME)
    x1 = pack(ids.point1.x, PRIME)
    y0 = pack(ids.point0.y, PRIME)

    value = new_x = (pow(slope, 2, SECP_P) - x0 - x1) % SECP_P
%}
```

```
fast_ec_add_assign_new_y:
%{ value = new_y = (slope * (x0 - new_x) - y0) % SECP_P %}
```

```
ec_mul_inner:
%{ value = new_y = (slope * (x - new_x) - y) % SECP_P %}
```

## Checklist
- [ ] Linked to Github Issue
- [x] Unit tests added
- [x] Integration tests added.
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.

### CODE REVIEW

File: src/vm/hints/secp/ec_utils.rs [hunk 1]
```diff
@@ -255,6 +258,151 @@ pub fn ec_double_assign_new_y(vm: &mut VirtualMachine) -> Result<(), VirtualMach
...
+    vm.exec_scopes
+        .assign_or_update_variable("y0", PyValueType::BigInt(y0));
+
+    vm.exec_scopes
+        .assign_or_update_variable("value", PyValueType::BigInt(value.clone()));
+
+    vm.exec_scopes
+        .assign_or_update_variable("new_x", PyValueType::BigInt(value));
+
+    Ok(())
```
Comment (Oppen) [inline src/vm/hints/secp/ec_utils.rs:353]:
```suggestion
    vm.exec_scopes
        .assign_or_update_variable("new_x", PyValueType::BigInt(value))
```

File: src/vm/hints/secp/ec_utils.rs [hunk 2]
```diff
@@ -255,6 +258,151 @@ pub fn ec_double_assign_new_y(vm: &mut VirtualMachine) -> Result<(), VirtualMach
...
+        vm.memory.get_integer(&(&point1_reloc + 1))?,
+        vm.memory.get_integer(&(&point1_reloc + 2))?,
+    );
+
+    let slope = pack(slope_d0, slope_d1, slope_d2, &vm.prime);
+    let x0 = pack(point0_x_d0, point0_x_d1, point0_x_d2, &vm.prime);
+    let x1 = pack(point1_x_d0, point1_x_d1, point1_x_d2, &vm.prime);
+    let y0 = pack(point0_y_d0, point0_y_d1, point0_y_d2, &vm.prime);
+
+    let value = (slope.pow(2).mod_floor(&SECP_P) - &x0 - x1).mod_floor(&SECP_P);
```
Comment (Oppen) [inline src/vm/hints/secp/ec_utils.rs:335]: Due to the equivalence in congruence of substraction, you could save one `mod_floor` by only performing it at the end:
```suggestion
    let value = (slope.pow(2) - &x0 - x1).mod_floor(&SECP_P);
```

File: src/vm/hints/secp/ec_utils.rs [hunk 3]
```diff
@@ -255,6 +258,151 @@ pub fn ec_double_assign_new_y(vm: &mut VirtualMachine) -> Result<(), VirtualMach
...
+    let scalar = get_integer_from_var_name(
+        "scalar",
+        ids,
+        &vm.memory,
+        &vm.references,
+        &vm.run_context,
+        hint_ap_tracking,
+    )?
+    .mod_floor(&vm.prime)
+    .mod_floor(&bigint!(2));
```
Comment (Oppen) [inline src/vm/hints/secp/ec_utils.rs:399]:
```suggestion
    .bit_and(1);
```

**Review by Oppen — APPROVED** [review]

File: src/vm/hints/secp/ec_utils.rs (hunk 2 above)
Comment (pefontana) [reply src/vm/hints/secp/ec_utils.rs]: Nice, I missed this.
Done

File: src/vm/hints/secp/ec_utils.rs (hunk 3 above)
Comment (pefontana) [reply src/vm/hints/secp/ec_utils.rs]: Done!

**Pull request merged**